Add feature to name a time series

GalaxyGorilla opened the pull request:
- add feature to name a series
- automatically ignore tags with value 'undefined'
- tags with key 'undefined' and the 'from_name' option
  remove an element from the name derived from the exometer id
- ignore 'undefined' elements in naming (when derived from
  exometer id)

GalaxyGorilla committed:
add series_name and formatting option

  THESE CHANGES ARE NOT BACKWARDS COMPATIBLE

  * add feature to name a series
  * add feature to remove tags which values are 'undefined'
  * add feature to remove tags which keys are 'undefined'
  * subscription options are now broken into
    - series_key
    - tags
    - formatting
  * removed traling whitespaces
  * updated README

surik commented:
Can you give some examples of transformation exometer ids to influxdb fields and tags?

GalaxyGorilla commented:
EDIT: I updated the PR and added tests. They show what happens here. I added two keywords in the whole PR: '$series_key' and '$delete'. The first is used to name the series, the second is used to remove an element from an exometer id (without creating a tag) when the name is derived from the id.

surik commented:
I'm thinking about ignoring `undefined` by default. What if someone wants to have a tag called `undefined`?
In my opinion this logic shouldn't be part of the reporter. We can get rid of `undefined` from name when we manipulate exometer_ids.

surik commented on test/exometer_influxdb_test.erl:
```diff
@@ -18,6 +18,34 @@ evaluate_subscription_tags_test() ->
+
+    ?assertEqual({[a,b], #{}},
+                 evaluate_subscription_tags([a, b, c], [{'$delete', {from_name, c}}])),
+
```
what if

``` erlang
evaluate_subscription_tags([a, b, c], [{'$delete', {from_name, 2}}, {tag, {from_name, 3}}])
```

?

GalaxyGorilla commented:
@surik I updated the PR according to our discussions. Since tags and formatting stuff are now splitted options there is no need anymore to use special keywords.

surik commented on README.md:
```diff
@@ -37,16 +37,19 @@ Available options:
 * __batch_window_size__ - set window size in ms for batch sending. This means reported will collect measurements within this interval and send all measurements in one packet. `0` by default. 
 
-Timestamping is by default done by influxdb itself. To enable `timestamping` with the reporter you can use `true` or `{true, Precision}` where `Precision` is a unit taken from `[n,u,ms,s,m,h]`. The default unit is `u`.
+The following options can be set globally in the application config or locally in a specific subscription. The latter case overwrites the first.
```
Why do we set it globally in application config? Can we set these like other options in reporter config?

GalaxyGorilla replied:
EDIT: Fixed.

hwinkel commented:
@surik the timestamp for a given metric should be set at the source. Asyou see with batching, unavailable backends etc etc. the receive timestamp of the backenend, in your case influx, will aways differ from the source timstamp. therefore both are important.

for the sample intervals the same is true. there is a difference how often a application updates the exo metrics push from app->exo. and how often the Reporter sends the stats out to the backend.

If we have the intended HTTP GET "reporter" discussed with @uwiger and @GalaxyGorilla the external GET (pull) interval will dictate the reporting interval. If you have a metric type of exometer_function this will also trigger the collection via the callback as far I understand Exometer correctly

surik commented on README.md:
````diff
@@ -56,14 +59,27 @@ Besides the `tags` for the reporter initialization it is possible to add other t
+This will result in a name `erlang` with the tag pair `{tag, memory}` (plus the default pair `{host, Host}`). To disable the removal of elements in the series name you can set:
+
+```
+{formatting, [{purge, [{all_from_name, false}]}]}
````
please set syntax to erlang

GalaxyGorilla replied:
Fixed

surik commented on README.md:
````diff
@@ -56,14 +59,27 @@ Besides the `tags` for the reporter initialization it is possible to add other t
+Further it might be handy to remove e.g. `undefined` tag keys or values. This can be done via:
+
+```
+{formatting, [{purge, [{tag_keys, undefined}, {tag_values, undefined}]}]}
````
please set syntax to erlang

GalaxyGorilla replied:
Fixed

surik commented:
@hwinkel I don't understand why you mentioned timestamp and bathing here. This PR is not about it.

hwinkel commented:
@surik moved timestamp handling and reporting interval discussion to a separate issue #14

GalaxyGorilla commented:
Updated and rebased.
